mapper/thin_edge_json/src/json.rs

```diff
@@ -282,12 +276,66 @@ impl ThinEdgeJsonError {
 
 mod tests {
     use super::*;
-
+    use crate::measurement::*;
     fn test_timestamp() -> DateTime<FixedOffset> {
         FixedOffset::east(5 * 3600)
             .ymd(2021, 04, 08)
             .and_hms(0, 0, 0)
     }
+    #[test]
+    fn test_str_with_invalid_time_stamp() {
+        let input = r#"{
+            "time" : "2013-06-2217:03:14.000658767+02:00"
+        }"#;
+        let expected_error = r#"Invalid ISO8601 timestamp (expected YYYY-MM-DDThh:mm:ss.sss.±hh:mm): "2013-06-2217:03:14.000658767+02:00": input contains invalid characters"#;
+        let output_err = ThinEdgeJson::from_str(input, test_timestamp()).unwrap_err();
+        assert_eq!(output_err.to_string(), expected_error);
+    }
+
+    #[test]
+    fn test_str_with_valid_timestamp() {
+        let input = r#"{
+            "time" : "2021-04-30T17:03:14+02:00",
+            "temperature" : 25
+        }"#;
+        let timestamp = FixedOffset::east(2 * 3600)
+            .ymd(2021, 04, 30)
+            .and_hms(17, 03, 14);
+
+        let output = ThinEdgeJson::from_str(input, timestamp).unwrap();
+        assert_eq!(output.timestamp, timestamp);
+    }
+    #[test]
+    fn test_str_with_millisecond_timestamp() {
+        let input = r#"{
+            "time" : "2021-04-30T17:03:14.123+02:00",
+            "temperature" : 25
+        }"#;
+        let timestamp = FixedOffset::east(2 * 3600)
+            .ymd(2021, 04, 30)
+            .and_hms_milli(17, 03, 14, 123);
+
+        let output = ThinEdgeJson::from_str(input, timestamp).unwrap();
+        assert_eq!(output.timestamp, timestamp);
+    }
+
+    #[test]
+    fn test_str_with_nanosecond_timestamp() {
+        let input = r#"{
+            "time" : "2021-04-30T17:03:14.123456789+02:00",
+            "temperature" : 25
+        }"#;
+
+        let timestamp = current_timestamp();

```

Comment:
where is `current_timestamp()` coming from??? I don't see the definition anywhere.

Please try to make test cases as self-containing as possible. For tests it's sometimes better to "repeat yourself" and copy and paste code in tests than trying to factor common code out into a shared function.

I think what you want here is:

```rust
let timestamp = FixedOffset::east(2 * 3600)
            .ymd(2021, 04, 30)
            .and_hms_milli(17, 03, 14, 123);
```

Reply:
Thanks for the hint, done.